// @kubiks/otel-autumn/dist/index.js
import { context, SpanKind, SpanStatusCode, trace, } from "@opentelemetry/api";
const DEFAULT_TRACER_NAME = "@kubiks/otel-autumn";
const INSTRUMENTED_FLAG = Symbol("kubiksOtelAutumnInstrumented");
// Semantic attribute constants
export const SEMATTRS_BILLING_SYSTEM = "billing.system";
export const SEMATTRS_BILLING_OPERATION = "billing.operation";
export const SEMATTRS_AUTUMN_RESOURCE = "autumn.resource";
export const SEMATTRS_AUTUMN_TARGET = "autumn.target";
// Customer attributes
export const SEMATTRS_AUTUMN_CUSTOMER_ID = "autumn.customer_id";
export const SEMATTRS_AUTUMN_ENTITY_ID = "autumn.entity_id";
// Product attributes
export const SEMATTRS_AUTUMN_PRODUCT_ID = "autumn.product_id";
export const SEMATTRS_AUTUMN_PRODUCT_IDS = "autumn.product_ids";
export const SEMATTRS_AUTUMN_PRODUCT_NAME = "autumn.product_name";
export const SEMATTRS_AUTUMN_PRODUCT_SCENARIO = "autumn.product_scenario";
// Feature attributes
export const SEMATTRS_AUTUMN_FEATURE_ID = "autumn.feature_id";
export const SEMATTRS_AUTUMN_FEATURE_NAME = "autumn.feature_name";
export const SEMATTRS_AUTUMN_ALLOWED = "autumn.allowed";
export const SEMATTRS_AUTUMN_BALANCE = "autumn.balance";
export const SEMATTRS_AUTUMN_USAGE = "autumn.usage";
export const SEMATTRS_AUTUMN_INCLUDED_USAGE = "autumn.included_usage";
export const SEMATTRS_AUTUMN_UNLIMITED = "autumn.unlimited";
export const SEMATTRS_AUTUMN_REQUIRED_BALANCE = "autumn.required_balance";
// Checkout attributes
export const SEMATTRS_AUTUMN_CHECKOUT_URL = "autumn.checkout_url";
export const SEMATTRS_AUTUMN_HAS_PRORATIONS = "autumn.has_prorations";
export const SEMATTRS_AUTUMN_TOTAL_AMOUNT = "autumn.total_amount";
export const SEMATTRS_AUTUMN_CURRENCY = "autumn.currency";
export const SEMATTRS_AUTUMN_FORCE_CHECKOUT = "autumn.force_checkout";
export const SEMATTRS_AUTUMN_INVOICE = "autumn.invoice";
// Track attributes
export const SEMATTRS_AUTUMN_EVENT_NAME = "autumn.event_name";
export const SEMATTRS_AUTUMN_VALUE = "autumn.value";
export const SEMATTRS_AUTUMN_EVENT_ID = "autumn.event_id";
export const SEMATTRS_AUTUMN_IDEMPOTENCY_KEY = "autumn.idempotency_key";
// Attach/Cancel attributes
export const SEMATTRS_AUTUMN_SUCCESS = "autumn.success";
function finalizeSpan(span, error) {
    if (error) {
        if (error instanceof Error) {
            span.recordException(error);
        }
        else {
            span.recordException(new Error(String(error)));
        }
        span.setStatus({ code: SpanStatusCode.ERROR });
    }
    else {
        span.setStatus({ code: SpanStatusCode.OK });
    }
    span.end();
}
function annotateCheckSpan(span, params) {
    span.setAttributes({
        [SEMATTRS_BILLING_SYSTEM]: "autumn",
        [SEMATTRS_BILLING_OPERATION]: "check",
        [SEMATTRS_AUTUMN_RESOURCE]: params.feature_id ? "features" : "products",
        [SEMATTRS_AUTUMN_TARGET]: params.feature_id
            ? "features.check"
            : "products.check",
    });
    span.setAttribute(SEMATTRS_AUTUMN_CUSTOMER_ID, params.customer_id);
    if (params.feature_id) {
        span.setAttribute(SEMATTRS_AUTUMN_FEATURE_ID, params.feature_id);
    }
    if (params.product_id) {
        span.setAttribute(SEMATTRS_AUTUMN_PRODUCT_ID, params.product_id);
    }
    if (params.entity_id) {
        span.setAttribute(SEMATTRS_AUTUMN_ENTITY_ID, params.entity_id);
    }
    if (typeof params.required_balance === "number") {
        span.setAttribute(SEMATTRS_AUTUMN_REQUIRED_BALANCE, params.required_balance);
    }
}
function annotateCheckResponse(span, response) {
    if (typeof response.allowed === "boolean") {
        span.setAttribute(SEMATTRS_AUTUMN_ALLOWED, response.allowed);
    }
    if (typeof response.balance === "number") {
        span.setAttribute(SEMATTRS_AUTUMN_BALANCE, response.balance);
    }
    if (typeof response.usage === "number") {
        span.setAttribute(SEMATTRS_AUTUMN_USAGE, response.usage);
    }
    if (typeof response.included_usage === "number") {
        span.setAttribute(SEMATTRS_AUTUMN_INCLUDED_USAGE, response.included_usage);
    }
    if (typeof response.unlimited === "boolean") {
        span.setAttribute(SEMATTRS_AUTUMN_UNLIMITED, response.unlimited);
    }
    if (typeof response.status === "string") {
        span.setAttribute(SEMATTRS_AUTUMN_PRODUCT_SCENARIO, response.status);
    }
}
function annotateTrackSpan(span, params) {
    span.setAttributes({
        [SEMATTRS_BILLING_SYSTEM]: "autumn",
        [SEMATTRS_BILLING_OPERATION]: "track",
        [SEMATTRS_AUTUMN_RESOURCE]: "events",
        [SEMATTRS_AUTUMN_TARGET]: "events.track",
    });
    span.setAttribute(SEMATTRS_AUTUMN_CUSTOMER_ID, params.customer_id);
    if (params.feature_id) {
        span.setAttribute(SEMATTRS_AUTUMN_FEATURE_ID, params.feature_id);
    }
    if (params.event_name) {
        span.setAttribute(SEMATTRS_AUTUMN_EVENT_NAME, params.event_name);
    }
    if (typeof params.value === "number") {
        span.setAttribute(SEMATTRS_AUTUMN_VALUE, params.value);
    }
    if (params.entity_id) {
        span.setAttribute(SEMATTRS_AUTUMN_ENTITY_ID, params.entity_id);
    }
    if (params.idempotency_key) {
        span.setAttribute(SEMATTRS_AUTUMN_IDEMPOTENCY_KEY, params.idempotency_key);
    }
}
function annotateTrackResponse(span, response) {
    if (response.id) {
        span.setAttribute(SEMATTRS_AUTUMN_EVENT_ID, response.id);
    }
}
function annotateCheckoutSpan(span, params, config) {
    span.setAttributes({
        [SEMATTRS_BILLING_SYSTEM]: "autumn",
        [SEMATTRS_BILLING_OPERATION]: "checkout",
        [SEMATTRS_AUTUMN_RESOURCE]: "checkout",
        [SEMATTRS_AUTUMN_TARGET]: "checkout.create",
    });
    span.setAttribute(SEMATTRS_AUTUMN_CUSTOMER_ID, params.customer_id);
    if (params.product_id) {
        span.setAttribute(SEMATTRS_AUTUMN_PRODUCT_ID, params.product_id);
    }
    if (params.product_ids && params.product_ids.length > 0) {
        span.setAttribute(SEMATTRS_AUTUMN_PRODUCT_IDS, params.product_ids.join(", "));
    }
    if (params.entity_id) {
        span.setAttribute(SEMATTRS_AUTUMN_ENTITY_ID, params.entity_id);
    }
    if (typeof params.force_checkout === "boolean") {
        span.setAttribute(SEMATTRS_AUTUMN_FORCE_CHECKOUT, params.force_checkout);
    }
    if (typeof params.invoice === "boolean") {
        span.setAttribute(SEMATTRS_AUTUMN_INVOICE, params.invoice);
    }
}
function annotateCheckoutResponse(span, response) {
    if (response.url) {
        span.setAttribute(SEMATTRS_AUTUMN_CHECKOUT_URL, response.url);
    }
    if (typeof response.has_prorations === "boolean") {
        span.setAttribute(SEMATTRS_AUTUMN_HAS_PRORATIONS, response.has_prorations);
    }
    if (typeof response.total === "number") {
        span.setAttribute(SEMATTRS_AUTUMN_TOTAL_AMOUNT, response.total);
    }
    if (response.currency) {
        span.setAttribute(SEMATTRS_AUTUMN_CURRENCY, response.currency);
    }
}
function annotateAttachSpan(span, params) {
    span.setAttributes({
        [SEMATTRS_BILLING_SYSTEM]: "autumn",
        [SEMATTRS_BILLING_OPERATION]: "attach",
        [SEMATTRS_AUTUMN_RESOURCE]: "products",
        [SEMATTRS_AUTUMN_TARGET]: "products.attach",
    });
    span.setAttribute(SEMATTRS_AUTUMN_CUSTOMER_ID, params.customer_id);
    span.setAttribute(SEMATTRS_AUTUMN_PRODUCT_ID, params.product_id);
    if (params.entity_id) {
        span.setAttribute(SEMATTRS_AUTUMN_ENTITY_ID, params.entity_id);
    }
}
function annotateAttachResponse(span, response) {
    if (typeof response.success === "boolean") {
        span.setAttribute(SEMATTRS_AUTUMN_SUCCESS, response.success);
    }
    if (response.checkout_url) {
        span.setAttribute(SEMATTRS_AUTUMN_CHECKOUT_URL, response.checkout_url);
    }
}
function annotateCancelSpan(span, params) {
    span.setAttributes({
        [SEMATTRS_BILLING_SYSTEM]: "autumn",
        [SEMATTRS_BILLING_OPERATION]: "cancel",
        [SEMATTRS_AUTUMN_RESOURCE]: "products",
        [SEMATTRS_AUTUMN_TARGET]: "products.cancel",
    });
    span.setAttribute(SEMATTRS_AUTUMN_CUSTOMER_ID, params.customer_id);
    span.setAttribute(SEMATTRS_AUTUMN_PRODUCT_ID, params.product_id);
}
function annotateCancelResponse(span, response) {
    if (typeof response.success === "boolean") {
        span.setAttribute(SEMATTRS_AUTUMN_SUCCESS, response.success);
    }
}
export function instrumentAutumn(client, config) {
    // Check if already instrumented
    if (client[INSTRUMENTED_FLAG]) {
        return client;
    }
    const tracer = trace.getTracer(DEFAULT_TRACER_NAME);
    // Instrument check method
    const originalCheck = client.check.bind(client);
    const instrumentedCheck = async function instrumentedCheck(params) {
        const span = tracer.startSpan("autumn.check", {
            kind: SpanKind.CLIENT,
        });
        annotateCheckSpan(span, params);
        const activeContext = trace.setSpan(context.active(), span);
        try {
            const result = await context.with(activeContext, () => originalCheck(params));
            if (result.data) {
                annotateCheckResponse(span, result.data);
            }
            finalizeSpan(span);
            return result;
        }
        catch (error) {
            finalizeSpan(span, error);
            throw error;
        }
    };
    // Instrument track method
    const originalTrack = client.track.bind(client);
    const instrumentedTrack = async function instrumentedTrack(params) {
        const span = tracer.startSpan("autumn.track", {
            kind: SpanKind.CLIENT,
        });
        annotateTrackSpan(span, params);
        const activeContext = trace.setSpan(context.active(), span);
        try {
            const result = await context.with(activeContext, () => originalTrack(params));
            if (result.data) {
                annotateTrackResponse(span, result.data);
            }
            finalizeSpan(span);
            return result;
        }
        catch (error) {
            finalizeSpan(span, error);
            throw error;
        }
    };
    // Instrument checkout method
    const originalCheckout = client.checkout.bind(client);
    const instrumentedCheckout = async function instrumentedCheckout(params) {
        const span = tracer.startSpan("autumn.checkout", {
            kind: SpanKind.CLIENT,
        });
        annotateCheckoutSpan(span, params, config);
        const activeContext = trace.setSpan(context.active(), span);
        try {
            const result = await context.with(activeContext, () => originalCheckout(params));
            if (result.data) {
                annotateCheckoutResponse(span, result.data);
            }
            finalizeSpan(span);
            return result;
        }
        catch (error) {
            finalizeSpan(span, error);
            throw error;
        }
    };
    // Instrument attach method
    const originalAttach = client.attach.bind(client);
    const instrumentedAttach = async function instrumentedAttach(params) {
        const span = tracer.startSpan("autumn.attach", {
            kind: SpanKind.CLIENT,
        });
        annotateAttachSpan(span, params);
        const activeContext = trace.setSpan(context.active(), span);
        try {
            const result = await context.with(activeContext, () => originalAttach(params));
            if (result.data) {
                annotateAttachResponse(span, result.data);
            }
            finalizeSpan(span);
            return result;
        }
        catch (error) {
            finalizeSpan(span, error);
            throw error;
        }
    };
    // Instrument cancel method
    const originalCancel = client.cancel.bind(client);
    const instrumentedCancel = async function instrumentedCancel(params) {
        const span = tracer.startSpan("autumn.cancel", {
            kind: SpanKind.CLIENT,
        });
        annotateCancelSpan(span, params);
        const activeContext = trace.setSpan(context.active(), span);
        try {
            const result = await context.with(activeContext, () => originalCancel(params));
            if (result.data) {
                annotateCancelResponse(span, result.data);
            }
            finalizeSpan(span);
            return result;
        }
        catch (error) {
            finalizeSpan(span, error);
            throw error;
        }
    };
    // Replace methods with instrumented versions
    client.check = instrumentedCheck;
    client.track = instrumentedTrack;
    client.checkout = instrumentedCheckout;
    client.attach = instrumentedAttach;
    client.cancel = instrumentedCancel;
    // Mark as instrumented
    client[INSTRUMENTED_FLAG] = true;
    return client;
}
