// autumn-js/dist/libraries/backend/chunk-56DMVCCC.mjs
import {
  addReferralRoutes
} from "./chunk-7MOIXHFG.mjs";
import {
  addEntityRoutes
} from "./chunk-DKXKK76P.mjs";
import {
  addGenRoutes
} from "./chunk-N3VJUVPG.mjs";
import {
  addProductRoutes
} from "./chunk-IVBYU3TD.mjs";
import {
  withAuth
} from "./chunk-RN3IUPE5.mjs";
import {
  fetchPricingTable
} from "./chunk-AMSY36CS.mjs";
import {
  BASE_PATH
} from "./chunk-KSG3E4Q2.mjs";

// src/libraries/backend/routes/backendRouter.ts
import { addRoute, createRouter } from "rou3";
var sanitizeCustomerBody = (body) => {
  let bodyCopy = { ...body };
  delete bodyCopy.id;
  delete bodyCopy.name;
  delete bodyCopy.email;
  return bodyCopy;
};
var createCustomerHandler = withAuth({
  fn: async ({
    autumn,
    customer_id,
    customer_data = {},
    body
  }) => {
    let res = await autumn.customers.create({
      id: customer_id,
      ...customer_data,
      ...sanitizeCustomerBody(body)
    });
    return res;
  }
});
var getPricingTableHandler = withAuth({
  fn: async ({
    autumn,
    customer_id
  }) => {
    return await fetchPricingTable({
      instance: autumn,
      params: {
        customer_id: customer_id || void 0
      }
    });
  },
  requireCustomer: false
});
var createRouterWithOptions = () => {
  const router = createRouter();
  addRoute(router, "POST", `${BASE_PATH}/cors`, {
    handler: () => {
      return {
        body: {
          message: "OK"
        },
        statusCode: 200
      };
    }
  });
  addRoute(router, "POST", `${BASE_PATH}/customers`, {
    handler: createCustomerHandler
  });
  addRoute(router, "GET", `${BASE_PATH}/components/pricing_table`, {
    handler: getPricingTableHandler,
    requireCustomer: false
  });
  addGenRoutes(router);
  addEntityRoutes(router);
  addReferralRoutes(router);
  addProductRoutes(router);
  return router;
};

export {
  createRouterWithOptions
};
