// autumn-js/dist/libraries/backend/chunk-DKXKK76P.mjs
import {
  withAuth
} from "./chunk-RN3IUPE5.mjs";

// src/libraries/backend/routes/entityRoutes.ts
import { addRoute } from "rou3";
var createEntityHandler = withAuth({
  fn: async ({
    autumn,
    customer_id,
    body
  }) => {
    return await autumn.entities.create(customer_id, body);
  }
});
var getEntityHandler = withAuth({
  fn: async ({
    autumn,
    customer_id,
    pathParams,
    searchParams
  }) => {
    if (!pathParams?.entityId) {
      return {
        statusCode: 400,
        body: {
          error: "no_entity_id",
          message: "Entity ID is required"
        }
      };
    }
    let params = {
      expand: searchParams?.expand?.split(",")
    };
    let res = await autumn.entities.get(
      customer_id,
      pathParams.entityId,
      params
    );
    return res;
  }
});
var deleteEntityHandler = withAuth({
  fn: async ({
    autumn,
    customer_id,
    pathParams
  }) => {
    if (!pathParams?.entityId) {
      return {
        statusCode: 400,
        body: {
          error: "no_entity_id",
          message: "Entity ID is required"
        }
      };
    }
    return await autumn.entities.delete(customer_id, pathParams.entityId);
  }
});
var addEntityRoutes = async (router) => {
  addRoute(router, "POST", "/api/autumn/entities", {
    handler: createEntityHandler
  });
  addRoute(router, "GET", "/api/autumn/entities/:entityId", {
    handler: getEntityHandler
  });
  addRoute(router, "DELETE", "/api/autumn/entities/:entityId", {
    handler: deleteEntityHandler
  });
};

export {
  addEntityRoutes
};
